### DISCUSSION TITLE
How to upsert multiple records at the same time?

### DISCUSSION THREAD

**Discussion by pedrommcarrasco** [groue/GRDB.swift#1624] [Q&A] [answered]
I have an array of records that I need to insert or update efficiently. I couldn’t find a Batch Upsert option in the documentation — could you provide any guidance?

**Comment by groue** [comment, marked as answer]
Hello @pedrommcarrasco,

Write a loop, and make sure this loop is wrapped a single write transaction:

```swift
// CORRECT: one loop wrapped in a single transaction.
try dbQueue.write { db in
    for player in players {
        try player.upsert(db)
    }
}

// VERY BAD PERFORMANCE: one loop performs many transactions.
// This forces a lot of slow file-system synchronizations.
for player in players {
    try dbQueue.write { db in
        try player.upsert(db)
    }
}
```

If you are surprised, please check [Many Small Queries Are Efficient In SQLite](https://sqlite.org/np1queryprob.html).

---

Now, if you are looking for the absolute best performance you can get with GRDB, then go down to the SQLite metal and use a prepared statement. The library does not provide any convenience api for building one[^1], so you'll write something like:

```swift
// BEST PERFORMANCE
func upsertPlayers(_ db: Database, _ players: [Player]) throws {
    let statement = try db.cachedStatement(sql: """
        INSERT INTO player (id, name, score)
        VALUES (?, ?, ?)
        ON CONFLICT DO UPDATE SET
          -- one line per upserted column
          name = excluded.name,
          score = excluded.score
        """)
     
     for player in players {
         try statement.execute(arguments: [
             player.id,
             player.name,
             player.score,
         ])
     }
}
```

See [UPSERT](https://www.sqlite.org/lang_upsert.html) for more information.

[^1]: An application should be able to do it from the [schema inspection methods](https://swiftpackageindex.com/groue/grdb.swift/documentation/grdb/databaseschema#querying-the-database-schema)

**Reply by pedrommcarrasco** [reply]
Love it - thank you!
